### ISSUE TITLE
Diversity Binning Scheme based on level traversal path

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#737]
I have an idea for a new binning scheme that can apply to any game where an A* agent (or any agent) charts a path through the level. The idea is to have a separate bin for each unique path, except not exactly because this would be too many bins, and because this would be directly tied to fitness in many cases which is often some kind of measurement of the path length. Proper diversity demands that fitness can vary within the bin. 

So, in order to reduce the number of bins and allow a bit more flexibility, instead of discretizing a level down to individual tiles, it can be discretized at a coarser level, such as 2x2 batches of tiles, or 3x3, etc (it can be parameterizable). The actual level paths can be taken, and mapped onto this coarse representation to figure out the coarse path. There will of course be multiple possible paths that share the same coarse path.

Basically, the in-order sequence of coarse cell locations (coordinates) can be appended into a String, and these Strings can be the keys of the HashMap representing the archive. This approach would hopefully result in a wide variety of levels that traverse the possible space in a variety of ways.

The only issue I foresee is that parts of the level not traversed by the A* path could really contain anything, and the random contents of these sections would not be subject to any kind of selection, since they would neither influence fitness nor diversity. I'm not sure that would actually be a problem though, and I think it is still worth implementing something first and then seeing how it turns out.

**Comment by schrum2** [comment]
@MBatt1 and I sat thinking about this a lot and I'm afraid it is infeasible. First, we were troubled by the flexibility of the representation, since the path could loop back on itself and have arbitrary length. Ultimately, the MAP Elites archive (at least in my implementation) has a have a fixed, known size.

So, we considered what would happen if we divided the level into coarse tiles and only considered whether or not a tile was ever entered (ignoring order). This is when we realized that for N coarse tiles, we would need 2^N bins (every possible binary sequence of length N represents a possible bin). However, a Java array cannot be bigger than 2^31 (max int size), meaning that a level could not be split into more than 31 coarse tiles. This is TOO coarse! We would not be able to measure any meaningful diversity in the level.

Currently, we are considering alternatives that use the level traversal path in other ways. Directly using the action sequence is a possibility.

Ultimately, we need to do one of two things. Either:
1) Come up with a binning scheme that requires fewer than 2^31 bins
2) Resort to some sort of storage mechanism that can support more than 2^31 bins, since as a database. This second option would require a complete reimplmentation of MAP-Elites, and might introduce performance issues.

We'll need to think on this issue a bit to decide if it is worth keeping open.

**Comment by schrum2** [comment]
@MBatt1 Actually, I think we can still use this binning scheme with Zelda in an interesting way. Basically, we can make each room of the dungeon be a grid cell. With a 5 by 5 dungeon that is 25 rooms, so there are 2^25 ways for have visited or not visited the rooms. This archive size is big, but should still fit in memory.

In the long run, we may also be able to find a systematic way to exclude certain combinations that are impossible. Specifically, it is impossible to visit a collection of rooms that are not connected, so we don't even have to have bins for those possibilities in the archive ... but I'm not sure what a systematic way to count only the continuous combinations is though. I'll need to think about it. In the meantime, just make the archive have a size of 2^(W*H) where W and H are the Width and Height of the dungeon in terms of rooms

**Comment by MBatt1** [comment]
Currently the bin labels theoretically have everything they need to run, however I have some suspicions that the way the bin labels are created and the way we get an insertion index may be slightly incorrect, so it needs some unit tests for now, then maybe an actual test once the methods are verified working

**Comment by MBatt1** [comment]
I committed some changes that I made to fix the bin generation to use the recursive solution, as well as adding the test case. I'm a bit confused as to where exactly I'm going wrong, however the test case generates a dungeon, creates a search, however it only makes 4 moves, all within a single room it seems. @schrum2 if you have any ideas how the Zelda dungeons are traversed or what I might be missing, something is definitely going wrong.

**Comment by schrum2** [comment]
Some progress was made today by actually visualizing the generated levels. What is the current state of this?

**Comment by MBatt1** [comment]
@schrum2 
test case is at LevelTraversalPathBinLabelsTest

**Comment by schrum2** [comment]
dev_maxx has been merged into dev, and will be deleted soon. So, this issue can be investigated further in dev

**Comment by schrum2** [comment]
When I run LevelTraversalPathBinLabelsTest with the viewDungeon code uncommented I get this:
![image](https://user-images.githubusercontent.com/8397236/163323173-e121f626-96be-4bdc-9010-618c80a449a6.png)
If I run the test with the code uncommented, then the test simply passes, although there are some Python errors at the end from the GAN script being improperly terminated.

@MBatt1 can you remind me what the issue here was? Does it not actually work despite the positive test result?

**Comment by MBatt1** [comment]
You have to press enter in the console to clear out the dungeon view and continue executing the code, I based this view function on one from the regular zelda code and it requires you to send an input to unpause the execution when viewing the level itself.

The grid should print out 1s and 0s in the console if I recall so that you can compare seeing the bitstring with the associated dungeon

**Comment by schrum2** [comment]
It looks like it works, but the question now is whether we can evolve with it using MAP Elites

**Comment by schrum2** [comment]
After making  a batch file and fixing some issues related to #740 I think this is actually working, at least for 4 by 4 dungeons. Will post some results after initial testing

**Comment by schrum2** [comment]
Ok, so this eventually terminated with an OutOfMemoryError, which isn't surprising. Here is the stack trace:
```
Exception in thread "main" java.lang.OutOfMemoryError: GC overhead limit exceeded
        at java.util.Arrays.copyOf(Arrays.java:3181)
        at java.util.ArrayList.grow(ArrayList.java:267)
        at java.util.ArrayList.ensureExplicitCapacity(ArrayList.java:241)
        at java.util.ArrayList.ensureCapacityInternal(ArrayList.java:233)
        at java.util.ArrayList.add(ArrayList.java:464)
        at edu.southwestern.util.search.GraphSearch.search(GraphSearch.java:53)
        at edu.southwestern.tasks.zelda.ZeldaDungeonTask.evaluate(ZeldaDungeonTask.java:174)
        at edu.southwestern.evolution.mapelites.MAPElites.newIndividual(MAPElites.java:525)
        at edu.southwestern.experiment.evolution.SteadyStateExperiment.run(SteadyStateExperiment.java:44)
        at edu.southwestern.MMNEAT.MMNEAT.run(MMNEAT.java:533)
        at edu.southwestern.MMNEAT.MMNEAT.evolutionaryRun(MMNEAT.java:762)
        at edu.southwestern.MMNEAT.MMNEAT.main(MMNEAT.java:722)
```
This was still fairly early in evolution, so we'll have to invent some major tricks to make this approach possible within a realistic amount of memory. Here are the results I had though.

Note: These are from running `batch\ZeldaLevelTraversalPath\ZeldaMAPElites-LevelTraversalPath-Direct.bat`

![image](https://user-images.githubusercontent.com/8397236/166575304-6200ea10-8012-4c0c-adb1-4969cdf99595.png)
![image](https://user-images.githubusercontent.com/8397236/166575599-de6541b2-7f03-4153-9509-08eca73d227b.png)

These indicate that new bins were constantly being discovered, but that a very small percentage of the total bins were occupied. However, we know that the memory space was still reserved for the empty bins. Additionally, we know that many of the bins are impossible to fill since they represent a disconnected graph (Link can't teleport around the dungeon).

**Comment by schrum2** [comment]
Here are some example levels that traverse many rooms:
![1111111111110100-0 76923-9116-solution](https://user-images.githubusercontent.com/8397236/166577789-56c066b0-f9fb-4212-ad72-450ea4dd8c34.png)
![1111111111110010-0 84615-8021-solution](https://user-images.githubusercontent.com/8397236/166577852-ba94d79d-a8c7-4c50-8acf-6bd78a1d2946.png)
The method seems promising if we can overcome the technical hurdles

**Comment by MBatt1** [comment]
Just some brief searching brought up fastutil [BigLists](https://fastutil.di.unimi.it/docs/it/unimi/dsi/fastutil/BigList.html), which are semantically the same but have 64 bits to work with, although I perhaps don't really see where the problem is, I'm not sure what part of this would be causing the garbage collector memory issue

**Comment by schrum2** [comment]
Well, one possible problem is that we're storing a giant array that is mostly empty. That does take up space, but because those empty slots are null references they are least are not taking up any space on the heap (since no actual level is being pointed to).  We're also storing a giant list of String labels, most of which are not possible.

According to the results above there were about 1300 distinct dungeons at the point of termination. However, in the original CPPN2GAN paper the CPPN2GAN archive for Zelda had about 1750 dungeons, and those were 10 x 10 room dungeons (100 rooms!): 

Original CPPN2GAN paper: https://collections.southwestern.edu/s/suscholar/media/596

So, I think changing how we define the archive and the bin labels to be more efficient could buy us the space we need.

**Comment by schrum2** [comment]
@alejmedinajr This is a challenging issue that will require some discussion. @MBatt1 was working on it recently and made some good progress, and you may reach out to him with questions about the code at some point, but for the most part you and I need to brainstorm on ways of using less memory. Your knowledge of Graph Theory could be helpful here. Start by reading the history of this thread and we'll talk more later.

**Comment by schrum2** [comment]
To get some more insight on this problem, we need to look at what sort of reduction in size we get when we restrict ourselves to connected graphs. For example, we know that with 2 by 2 dungeons (4 rooms) there are 2^4 = 16 possible binary strings. It's easy to list them all out: 0000, 0001, 0010, 0011, ... , 1111. The question is, which ones are disconnected and should be disallowed? Is there a pattern? Write out (or compute with code) the list of all connected binary sequences for 3 by 3 and 4 by 4 dungeons, and see if there is a pattern.

**Comment by alejmedinajr** [comment]
I started looking into how many dungeons are disconnected. In terms of 2 by 2 dungeons, it appears that only 3 are disconnected. This is not too much of a surprise though since the only way to have disconnected rooms in a 2 by 2 is with the diagonals having the value of 1, or there being no 1's present at all (0000). I did look into the 3 by 3 dungeons as well, and these were more interesting. There seems to be more than 50% disconnected graphs out of 512 (2^9) total possible graphs. I have recorded all of this in a google sheet, and this sheet is uploaded into the SCOPE Artifacts folder in a subfolder called: SCOPE 2022 Diversity Binning Scheme based on level traversal path.
![Issue737 supporting image from google sheet](https://user-images.githubusercontent.com/100097809/168347360-7c250c62-c3c3-406e-af62-5872054432fc.png)

**Comment by alejmedinajr** [comment]
After looking at around 1,000 4 by 4's, it does seem like there will be a greater quantity of disconnected graphs when compared to the 2 by 2's or 3 by 3's. Although 1,000 is nowhere near the total possible graphs (65,536), I do not see the ratio of disconnected graphs drastically decreasing.

**Comment by MBatt1** [comment]
I just had an interesting discussion with some coworkers about optimizing this...

What if instead of storing things as a bitstring of the 4x4 dungeon space, if it were stored as the shape of the dungeon room path itself. An example of this could be reducing the bitstrings

0000        1100        0000        
1100        0110        0000        
0110        0000        0110        
0000        0000        0011        

all into the same shape of

110
011

Thus a pattern could be to brute force all possible 1x1, 2x2, 2x3, 3x2, etc into their possible SHAPES rather than discretizing based on overall 4x4 grid placement. I'm not sure this attempt is especially useful/efficient, but it could be something worth thinking about?

**Comment by schrum2** [comment]
@MBatt1 Could you elaborate a bit? I'm not understanding the mapping, partly because the examples you posted wouldn't correspond to a 4 x 4 room; there would need to be 16 bits in each bit string, but these examples only have 12 each.

**Comment by MBatt1** [comment]
Ah looks like github didn't like my formatting, here's what I meant
![image](https://user-images.githubusercontent.com/53229958/168411366-d5ad5824-3474-4b07-bdc7-527701a514f2.png)

**Comment by schrum2** [comment]
Got it. I thought you were presenting 4 dungeons (each row) rather than 3 dungeons (each 4x4 square). This is a good idea, but I think it's a separate issue, so I've created issue #757.

@alejmedinajr should implement the original idea of this issue first. In fact, just to make sure this is documented, the plan for what to do next is to create a HashMap when the Binning Scheme is initiated that indicates which binary string maps to each index in the binning scheme. The reason this HashMap is needed is that the labels will no longer contain bit strings representing disconnected (and thus impossible) dungeons. However, this makes it hard to know which index each bit string will be at, hence the need for the HashMap

**Comment by alejmedinajr** [comment]
After my latest commit related to this issue, I ran the batch file in a power shell. It has been running for about an hour without crashing, so I will leave it running overnight (hopefully), and we will see what will happen then.

**Comment by schrum2** [comment]
Ok, I ran the code until it crashed, and unfortunately is seemed to crash around the same point it had before. Here is a plot of the number of bins filled:
![image](https://user-images.githubusercontent.com/8397236/168862920-54a7c7d0-3ebf-4da4-ae02-0b349034d2a9.png)
This actually filled fewer bins before crashing than the earlier results posted in this thread. We definitely need more analysis to see what is going on.

Start by reporting the number of bins that exist with the new scheme as proposed in the previous comment on this thread.

**Comment by alejmedinajr** [comment]
**Without** removing the superfluous labels, a 4 x 4 dungeon would have 2^16 or **65,536** possible labels. **With** removing the superfluous labels, the size of labels is 680 (when doing a maven install), which does not seem right since that is ~1%.

![issue737Analysis](https://user-images.githubusercontent.com/100097809/168911977-fcd26c45-b56a-4c4f-b23e-070de0e1d80b.png)

**Comment by schrum2** [comment]
I was looking at the LevelTraversalPathBinLabels class and noticed some outstanding TODO items that could make an impact.

Specifically, we currently set the labels like this:
`labels = new ArrayList<String>(numOfBins); // TODO: Change to make smaller`
I removed the `numOfBins` to have a small initial size. I also found the `trimToSize` method that cuts back the ArrayList capacity to the minimal amount, after all the labels have been created.

Testing now

**Comment by schrum2** [comment]
Actually, I ran the test batch file with an added print statement, and I saw this:
`Number of LevelTraversalPathBinLabels: 65536`
This should be easy to fix. The next commit shows what line of code to move. However, I already tried this and it broke a unit test, so I'm going to leave it to you to figure out.

**Comment by alejmedinajr** [comment]
Latest commit fixes the unit tests. The failure was caused by a combination of having some of the old approach methods not being accurate with the new approach bin scheme. There was also an error with the new approach and how it determined how a location was in bounds.

**Comment by schrum2** [comment]
Despite the fixes, and the lower number of bins in the new representation (11507 instead of 65536 which is about 17%) we are still getting an out of memory error at about the same point.

I fear the only way to make progress on this issue at this point is to use the memory profiler, which is a bit of a hassle. Put this off for a while.
